Add file based system access control

amrutagokhale opened the pull request:
Main motivation for this feature is to add the ability to have access control on catalogs so that administrators can restrict access to certain catalogs to certain users. This can only be done at the system level and cannot be achieved at the connector level. The file based system access control allows you to specify catalog access control rules in a file, similar to the connector level file based access control. 

The current implementation handles only the catalog access control rules. This can be extended in future to allow other types of rules.   

TD review here: https://github.com/Teradata/presto/pull/516

cawallin commented on presto-main/src/main/java/com/facebook/presto/security/AccessControl.java:
```diff
@@ -37,6 +37,11 @@
     /**
+     * Check whether identity is allowed to access catalog
+     */
+    void checkCanAccessCatalog(Identity identity, String catalogName);
```
All the checkCanAccessCatalog changes belong in https://github.com/prestodb/presto/pull/7695/commits/30730672acfcc5e73d4456ae05ac6394a1a7ec73, this commit should just be adding the file based system access control.

cawallin commented on presto-main/src/main/java/com/facebook/presto/security/AccessControlManager.java:
```diff
@@ -157,11 +160,21 @@ public void checkCanSetUser(Principal principal, String userName)
         requireNonNull(identity, "identity is null");
         requireNonNull(schemaName, "schemaName is null");
 
+        authenticationCheck(() -> checkCanAccessCatalog(identity, schemaName.getCatalogName()));
```
Could you check catalog access when you get the catalog handle? e.g. MetadataManager.getCatalogHandle?

amrutagokhale replied:
Good point. I looked into this. I am not sure what all such methods are, that need to have the catalog access control check, if we move it out of `AccessControlManager`. For example, even if I add a check in both, `getCatalogHandle` and `getTableHandle`, there are still other statements that would be missed since they call none of these two methods (e.g. methods in `MetadataListing.java`).

dain commented on presto-main/src/main/java/com/facebook/presto/security/FileBasedSystemAccessControl.java:
```diff
@@ -0,0 +1,238 @@
+            String configFileName = config.get(CONFIG_FILE_NAME);
+            checkState(
+                    configFileName != null,
+                    "Configuration option %s not found while loading file based system access control", CONFIG_FILE_NAME);
```
Maybe be more direct... like, `Security configuration must contain the '%s' property`

dain commented on presto-main/src/main/java/com/facebook/presto/security/FileBasedSystemAccessControl.java:
```diff
@@ -0,0 +1,238 @@
+                if (!path.isAbsolute()) {
+                    path = path.toAbsolutePath();
+                }
+                ImmutableList.Builder<CatalogAccessControlRule> catalogRulesBuilder = ImmutableList.builder();
```
I would check that the path is a file you `canRead` and throw an explicit exception

dain commented on presto-main/src/main/java/com/facebook/presto/security/FileBasedSystemAccessControlRules.java:
```diff
@@ -0,0 +1,38 @@
+    private final List<CatalogAccessControlRule> catalogRules;
+
+    @JsonCreator
+    public FileBasedSystemAccessControlRules(
```
no need to wrap this

dain commented on presto-main/src/main/java/com/facebook/presto/security/FileBasedSystemAccessControlRules.java:
```diff
@@ -0,0 +1,38 @@
+    public FileBasedSystemAccessControlRules(
+            @JsonProperty("catalogs") Optional<List<CatalogAccessControlRule>> catalogRules)
+    {
+        this.catalogRules = catalogRules.orElse(ImmutableList.of());
```
```
        this.catalogRules = catalogRules.map(ImmutableList::copyOf).orElse(ImmutableList.of());
```

dain commented on presto-main/src/test/java/com/facebook/presto/security/TestFileBasedSystemAccessControl.java:
```diff
@@ -0,0 +1,162 @@
+
+public class TestFileBasedSystemAccessControl
+{
+    private static final Principal PRINCIPAL = new TestingPrincipal("principal");
```
The principal isn't needed since it isn't used by the access control, so you can replace all uses with `Optional.empty`

dain commented on presto-main/src/test/java/com/facebook/presto/security/TestFileBasedSystemAccessControl.java:
```diff
@@ -0,0 +1,162 @@
+        AccessControlManager accessControlManager =  new AccessControlManager(transactionManager);
+
+        String path = this.getClass().getClassLoader().getResource("catalog.json").getPath();
+        accessControlManager.setSystemAccessControl(FileBasedSystemAccessControl.NAME, ImmutableMap.of(
```
no need to wrap this

dain commented on presto-main/src/test/java/com/facebook/presto/security/TestFileBasedSystemAccessControl.java:
```diff
@@ -0,0 +1,162 @@
+                    accessControlManager.checkCanGrantTablePrivilege(transactionId, alice, SELECT, aliceTable);
+                    accessControlManager.checkCanRevokeTablePrivilege(transactionId, alice, SELECT, aliceTable);
+                });
+        try {
```
use `Assert.assertThrows()` for all of these

amrutagokhale commented on presto-main/src/main/java/com/facebook/presto/security/AllowAllSystemAccessControl.java:
```diff
@@ -175,14 +174,4 @@ public void checkCanCreateViewWithSelectFromView(Identity identity, CatalogSchem
     }
-
-    @Override
-    public void checkCanGrantTablePrivilege(Identity identity, Privilege privilege, CatalogSchemaTableName table)
```
Thanks for the pointer. I commented on that PR.

amrutagokhale commented on presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java:
```diff
@@ -157,6 +157,9 @@ public Session createSession(
             SessionPropertyManager sessionPropertyManager)
     {
         accessControl.checkCanSetUser(identity.getPrincipal().orElse(null), identity.getUser());
+        if (catalog != null) {
```
I thought that we should have a check about whether a user can access the catalog before calling `setCatalog()` method. But I didn't see think of this case when a query may not be accessing the catalog. I have removed this code for now.

amrutagokhale commented:
@dain I've addressed your comments. 

Regarding reusing the code from `FileBasedAccessControl`, that's how I had started initially. Since that code is part of `presto-plugin-toolkit` module and I was trying to use it in `presto-main`, there were dependency issues that I couldn't resolve and didn't go that way. Also, that code has schema and table rules which wasn't needed for catalog access control. Let me know if you have a suggestion.

amrutagokhale commented:
I am going to look into Travis build failure.

amrutagokhale commented:
@dain I have fixed the relevant test failures. The tests that are failing now don't seem related.

I have kept the new commit as a separate `fixup` commit, so that it will be easier to review the new changes (in which I introduced the method `checkCanAccessCatalog` to SystemAccessControl interface).

dain commented:
Merged, thanks!
